chore: Refactor assertion logging

BlueWinds opened the pull request:
### User facing changelog
No change in behavior. Refactoring only.

### Additional details
As part of the Detached DOM work, I've run into a lot of issues with the way assertion logging currently works. This PR simplifies our use of the "onBeforeLog" hook, moving the logic for "is this the same log message" into easier to find and understand places.

### Steps to test
The automated tests are pretty thorough in this regard. Feel free to play around with any existing project, or especially `assertions.cy.ts` to verify that nothing has changed.

### How has the user experience changed?

### PR Tasks
- [x] Have tests been added/updated?
- [n/a] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

BlueWinds commented on packages/driver/cypress/e2e/commands/assertions.cy.js:
```diff
@@ -790,19 +789,18 @@ describe('src/cy/commands/assertions', () => {
     })
 
-    it('ends and snapshots immediately and sets child', (done) => {
+    it('snapshots immediately and sets child', (done) => {
```
This PR does introduce one test change - the assertion log is in a `pending` state when initially added, and is `passed` asynchronously as part of the `finishAssertions()` function, which happens slightly later in the lifecycle of the command.

Other tests verify that it moves from the `pending` state to `resolved` before the next command executes - it's just slightly later in resolving the same promise chain than `log:added`. I don't believe this is a meaningful difference, and so this can remain a chore even though I had to update one test. :woman_shrugging:

lmiller1990 commented on packages/driver/src/cy/commands/asserting.ts:
```diff
@@ -58,28 +96,38 @@ export default function (Commands, Cypress, cy, state) {
+              // we need to log out this .should since there
+              // was a problem with the actual assertion syntax
+              if (err.retry === false) {
+                return throwAndLogErr(err)
```
How can I get in here? I tried a few things, but not entirely clear how to reach this branch.

lmiller1990 replied:
Actually this one does the job: https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/e2e/commands/assertions.cy.js#L379-L384

BlueWinds replied:
I'm going to be moving this file around and refactoring it a bunch more as part of Detached DOM, so I didn't spend a ton of time commenting every detail. I think the upcoming changes should make this area a bit more straightforward, separating out "command errors" from "execution errors".

lmiller1990 commented on packages/driver/src/cypress/log.ts:
```diff
@@ -348,10 +348,10 @@ export class Log {
-      at: null,
-      next: null,
-    })
+    if (this.get('next')) {
```
Any context on why we do this? I know this wasn't changed as part of this PR, but when I removed it and ran the tests, only navigation.cy.ts failed, specifically " it.only('logs before + after', () => {". The test doesn't exactly shine much light on the actual reason we do this, though.

lmiller1990 replied:
Right, it's the label for `before/after`. Without it, you get the id:

![image](https://user-images.githubusercontent.com/19196536/184780399-83b01f60-15ca-4c08-bfc8-dc74e1314579.png)

BlueWinds replied:
Yep, just so. This was fairly arcane to me as well, escaped notice until the most recent commits.

lmiller1990 reviewed (commented):
All tests pass and I don't see any obvious problems, so looks good. Asked a few questions.

lmiller1990 commented:
Actually, this might be incorrect? I ran a spec in app - `cypress/e2e/cypress-in-cypress-e2e.cy.ts`. I ran the first test using `.only`. I'm seeing a lot of extra logs.

![image](https://user-images.githubusercontent.com/19196536/184780910-1ec4b8ec-d454-4fc8-8923-000327d1fc86.png)

lmiller1990 reviewed (changes requested):
Possible regression, going to request changes until we verify it's not a regression.

BlueWinds commented:
> Actually, this might be incorrect? I am seeing a lot of red in the command log.

> I ran a spec in app - cypress/e2e/cypress-in-cypress-e2e.cy.ts. I ran the first test using .only. I don't see this in develop on the same test. cc @BlueWinds

Oof, yeah, I see the same. Investigating, will fix and add another driver test once I figure out what's going on here. Wondering if it has to do with cypress-in-cypress.

BlueWinds commented:
> Possible regression, going to request changes until we verify it's not a regression.

Ah, figured it out. `.should('not.exist')` was getting duplicated when it retried; there's no assertion on the number of logs for any of the `not.exist` tests.

I've fixed the issue - the 'deduplicate logs' logic just needed to be moved slightly higher in the function - and added an assertion to an existing test to prevent future regressions in the area. See https://github.com/cypress-io/cypress/pull/23354/commits/6849134c4a4b27ba72d06e0da2112eae322361dd for the exact changes, and good catch!

lmiller1990 reviewed (approved):
Regression was fixed and comments responded too. I am not driver expert yet, so defs need that second 👀 , but seems like things behave the same as they did prior to and after this refactor 👍

emilyrohrbough commented on packages/driver/cypress/e2e/commands/assertions.cy.js:
```diff
@@ -790,19 +793,18 @@ describe('src/cy/commands/assertions', () => {
+          return
+        }
 
-          expect(log.get('ended')).to.be.true
-          expect(log.get('state')).to.eq('passed')
```
Are theses no longer true?

BlueWinds replied:
Correct. I put an explanation of the change here: https://github.com/cypress-io/cypress/pull/23354#discussion_r946210351

emilyrohrbough commented on packages/driver/src/cypress/log.ts:
```diff
@@ -624,13 +598,53 @@ class LogManager {
+
+      // if snapshot was passed
+      // in, go ahead and snapshot
+      if (log.get('snapshot') && !log.get('snapshots')) {
```
```suggestion
      if (log.get('snapshot')) {
```

BlueWinds replied:
Ah, yep. This was a remnant from earlier attempts to rein in the number of duplicate snapshots, but no longer necessary now that previousLogInstance logic is working correctly.

emilyrohrbough commented on packages/driver/src/cypress/log.ts:
```diff
@@ -624,13 +598,53 @@ class LogManager {
+        // Short version: If onBeforeLog sets the same commandLogId as an existing log on the current command,
+        // then we want to merge the current instance into the existing one, rather than display a new log to
+        // the user.
+        let commandLogId = log.get('commandLogId')
+
+        if (commandLogId != null) {
```
This logic is specific to `commands/asserting.ts` and commands are responsible for managing their own logs. Theres no need to move this logic here. This also puts additional attributes on the log that don't seem necessary which are stored on on the runner and the command.

flotwig reviewed (approved):
Nice!

emilyrohrbough commented on packages/driver/src/cypress/log.ts:
```diff
@@ -648,13 +622,24 @@ class LogManager {
 
-      // set the log on the command
-      const current = state('current')
+      let command = state('current')
```
```suggestion
      const command = state('current')
```

BlueWinds replied:
Applied.

emilyrohrbough commented on packages/driver/src/cypress/log.ts:
```diff
@@ -667,9 +652,8 @@ class LogManager {
-      // with no command reference, so just end the log
-      if (!current) {
+      // if the log isn't associated with a command, then we know it won't be retrying and we should just end it.
+      if (!state('current') || log.get('end')) {
```
```suggestion
      if (!command || log.get('end')) {
```

BlueWinds replied:
Applied.

emilyrohrbough commented on packages/driver/src/cypress/log.ts:
```diff
@@ -355,15 +355,7 @@ export class Log {
     this.set('snapshots', snapshots)
 
     if (options.next && shouldRebindSnapshotFn) {
-      const originalLogSnapshotFn = this.snapshot
-
-      this.snapshot = function () {
-        // restore the original snapshot function
-        this.snapshot = originalLogSnapshotFn
-
-        // call orig fn with next as name
-        return originalLogSnapshotFn.call(this, options.next)
-      }
```
We need to maintain this for cross-origin snapshots.

BlueWinds replied:
I merged in develop, which included a bunch of cross-origin changes, and fixed several merge conflicts; the tests were initially failing, but after some playing around I got them passing again.

Do you have a specific instance in mind where it's failing to snapshot cross-origin in this branch, so I can add a test?

emilyrohrbough replied:
@AtofStryker ^^

AtofStryker replied:
@emilyrohrbough @BlueWinds I don't believe there is an impact here to cross origin snapshots. 

In the case of asking the cross origin spec bridge for a snapshot, this should only happen in the case that something in the primary instance of cypress is requesting a snapshot while a `cy.origin` block is active or is about to be active, which means it's highly unlikely to be a command asking for the snapshot (ex: XHR's and fetch requests that ask for snapshots as a generated log that aren't direct command invocations). The tests in [snapshots.cy.ts](https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/e2e/e2e/origin/snapshots.cy.ts) should cover this for known cases. 

In the case were we DO need to rebind the original snapshot function, this usually happens on interaction events which should almost always be commands. Commands that are run within `cy.origin` take snapshots within the spec bridge itself, so no asking spec bridges for snapshots should be occuring. This should be covered by the various [command](https://github.com/cypress-io/cypress/tree/develop/packages/driver/cypress/e2e/e2e/origin/commands) tests that assert on `consoleProps` and other log attributes to verify the snapshot is correct.

Since we don't have tests that verify snapshots interactively (can't leverage the system test and cy-in-cy doesn't support snapshots), I ran this manually to double check and everything looks good from what I can see 😄 .

BlueWinds replied:
Thank you for double checking! I'd hate to have accidentally broken something you merged in just a few days ago.

AtofStryker replied:
Of course! Honestly wish we had a better way to test snapshots moving forward. Getting snapshot testing into `cy-in-cy` I think would cover a lot of ground.

emilyrohrbough commented on packages/driver/src/cy/commands/asserting.ts:
```diff
@@ -7,6 +7,30 @@ import $errUtils from '../../cypress/error_utils'
+const onBeforeLog = (log, command, commandLogId) => {
+  log.set('commandLogId', commandLogId)
+
+  if (command) {
```
When would command be undefined here?

BlueWinds replied:
It wouldn't be here. Remnant of when this was in log.ts, where there might not be a current command. Will remove the check.

BlueWinds commented on packages/driver/src/cy/commands/asserting.ts:
```diff
@@ -7,6 +7,30 @@ import $errUtils from '../../cypress/error_utils'
 const reHaveLength = /length/
 
+const onBeforeLog = (log, command, commandLogId) => {
+  log.set('commandLogId', commandLogId)
```
I tried to use the log id initially, you can't set a log's ID:

```
    // dont ever allow existing id's to be mutated
    if (this.attributes.id) {
      delete this.obj.id
    }
```

because the ID is meant to be *completely unique* - to identify a log instance, across the entire test. Updating it in the log would also require updating the mapping of id -> Log instance that the `LogManager` maintains, which begins to look like the same level of complexity as having its own attribute.

Also important to note that in the current implementation, the commandLogId isn't globally unique - it's unique (and reproducible) only within the context of a single command. Different instances of `.should()` can generate the same commandLogId. This isn't a problem in the current implementation, because we only search `command.get('logs')` for a matching commandLogId.

BlueWinds replied:
Basically, this just serves a different purpose than the log's id. It's a bit of state-tracking for an instance of `.should()`.

BlueWinds commented on packages/driver/src/cy/commands/asserting.ts:
```diff
@@ -7,6 +7,30 @@ import $errUtils from '../../cypress/error_utils'
+        previousLogInstance.end()
+      }
+
+      return false
```
Wrote a slightly different comment (with the same meaning).
